gasprice/feehistory.go

```diff
@@ -0,0 +1,141 @@
+package gasprice
+
+import (
+	"encoding/binary"
+	"errors"
+	"log"
+	"math"
+	"math/big"
+	"sort"
+)
+
+var (
+	ErrInvalidPercentile = errors.New("invalid percentile")
+	ErrBlockCount        = errors.New("blockCount must be greater than 0")
+	ErrBlockInfo         = errors.New("could not find block info")
+)
+
+const (
+	maxBlockRequest = 1024
+)
+
+type cacheKey struct {
+	number      uint64
+	percentiles string
+}
+
+// processedFees contains the results of a processed block.
+type processedFees struct {
+	reward       []uint64
+	baseFee      uint64
+	gasUsedRatio float64
+}
+
+type txGasAndReward struct {
+	gasUsed *big.Int
+	reward  *big.Int
+}
+
+func (g *GasHelper) FeeHistory(blockCount uint64, newestBlock uint64, rewardPercentiles []float64) (*uint64, *[]uint64, *[]float64, *[][]uint64, error) {
+	if blockCount < 1 {
+		return nil, nil, nil, nil, ErrBlockCount
+	}
+	if newestBlock > g.backend.Header().Number {
+		newestBlock = g.backend.Header().Number
+	}
+	if blockCount > maxBlockRequest {
+		blockCount = maxBlockRequest
+	}
+	if blockCount > newestBlock {
+		blockCount = newestBlock
+	}
+	for i, p := range rewardPercentiles {
+		if p < 0 || p > 100 {
+			return nil, nil, nil, nil, ErrInvalidPercentile
+		}
+		if i > 0 && p < rewardPercentiles[i-1] {
+			return nil, nil, nil, nil, ErrInvalidPercentile
+		}
+	}
+
+	var (
+		oldestBlock   = newestBlock - blockCount + 1
+		baseFeePerGas = make([]uint64, blockCount+1)
+		gasUsedRatio  = make([]float64, blockCount)
+		reward        = make([][]uint64, blockCount)
+	)
+	if oldestBlock < 1 {
+		oldestBlock = 1
+	}
+	percentileKey := make([]byte, 8*len(rewardPercentiles))
+	for i, p := range rewardPercentiles {
+		binary.LittleEndian.PutUint64(percentileKey[i*8:(i+1)*8], math.Float64bits(p))
+	}
+
+	for i := oldestBlock; i < oldestBlock+blockCount; i++ {
```

Comment:
I think we can just use `i <= newestBlock` instead of `i < oldestBlock+blockCount`.